server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRef.java

```diff
@@ -0,0 +1,155 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.qualityprofile;
+
+import javax.annotation.Nullable;
+import org.sonar.api.resources.Languages;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.server.component.ws.LanguageParamUtils;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+import static org.apache.commons.lang.StringUtils.isEmpty;
+
+/**
+ * Reference to a Quality profile. The two exclusive options to reference a profile are:
+ * <ul>
+ *   <li>by its key</li>
+ *   <li>by the couple {language, name}</li>
+ * </ul>
+ */
+public class QProfileRef {
+
+  public static final String PARAM_LANGUAGE = "language";
+  public static final String PARAM_PROFILE_NAME = "profileName";
+  public static final String PARAM_PROFILE_KEY = "profileKey";
+
+  private final String key;
+  private final String language;
+  private final String name;
+
+  private QProfileRef(@Nullable String key, @Nullable String language, @Nullable String name) {
+    this.key = key;
+    this.language = language;
+    this.name = name;
+  }
+
+  /**
+   * @return {@code true} if key is defined and {@link #getKey()} can be called. If {@code false}, then
+   *   the couple {language, name} is defined and the methods {@link #getLanguage()}/{@link #getName()}
+   *   can be called.
+   */
+  public boolean hasKey() {
+    return this.key != null;
+  }
+
+  /**
+   * @return non-null key
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code true}
+   */
+  public String getKey() {
+    checkState(key != null, "Key is not present. Please call hasKey().");
+    return key;
+  }
+
+  /**
+   * @return non-null language
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code false}
+   */
+  public String getLanguage() {
+    checkState(language != null, "Language is not present. Please call hasKey().");
+    return language;
+  }
+
+  /**
+   * @return non-null name
+   * @throws IllegalStateException if {@link #hasKey()} does not return {@code false}
+   */
+  public String getName() {
+    checkState(name != null, "Name is not present. Please call hasKey().");
+    return name;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    QProfileRef that = (QProfileRef) o;
+    if (key != null ? !key.equals(that.key) : (that.key != null)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS864)